### ISSUE TITLE
More helpful logs when service "Create Failed"

### ISSUE THREAD

**Issue by kdhillon** [aws/apprunner-roadmap#110]
**Tell us about your request**
When creating a new service, create fails without any errors in deployment logs. The last visible logs are 
```
02-14-2022 10:18:10 PM [Build] Successfully tagged application-image:latest
02-14-2022 10:18:10 PM [Build] Successfully built 2a53f3e899cb
```
Which suggests that build succeeded. Seems like there should always be an error logged in deployment log anytime a Service Create fails.

**Describe alternatives you've considered**
Tried looking through forums and past posts, all of them suggest checking logs, but event logs and/or deployment logs show no errors explaining why it failed.

**Additional context**
ARN: arn:aws:apprunner:us-west-2:520917683465:service/arpeggi-web-beta2/600285cea79a48fbb137786c78d31b52
Another ARN: arn:aws:apprunner:us-west-2:520917683465:service/arpeggi-web-beta/ea870ffb4d0a44979811a77ac1bf96f5

Current Service Quota: 10. 
Existing Services: 3.

Repro steps (reproducible 5/5 tries): 
- "Create Service"
- "Source code repository"
- Connect github repo + branch (we have another Service correctly set up and running from the exact same github branch)
- Deployment settings: Manual or Automatic (I've tried both)
- Configure settings to be exact same as existing Service
- Default settings (AWS managed key, etc)
- Wait for "Create service in progress..." to complete, always get "Create Failed"
- Event log is always:
```
02-14-2022 10:08:30 PM [AppRunner] Starting source code build.
02-14-2022 10:08:29 PM [AppRunner] Successfully pulled source code.
02-14-2022 10:07:41 PM [AppRunner] Service status is set to OPERATION_IN_PROGRESS.
02-14-2022 10:07:41 PM [AppRunner] Service creation started.
```

Deployment Log for "Create Service" contains no errors:
https://us-west-2.console.aws.amazon.com/cloudwatch/home?region=us-west-2#logsV2:log-groups/log-group/$252Faws$252Fapprunner$252Farpeggi-web-beta2$252F600285cea79a48fbb137786c78d31b52$252Fservice/log-events/deployment$252F1ecf04a9c05c4bf1a7410d61a652aff6

**Comment by tsheaff** [comment]
Yeah this is crazy. I've run into the same issue. I have absolutely no way to debug why it's failing!

**Comment by tsheaff** [comment]
I'm sure it's an easily solvable issue for why the creation is failing, but with no way to inspect what went wrong, all I can do is poke around in the dark for hours 😠

**Comment by tsheaff** [comment]
cc @akshayram-wolverine

**Comment by tsheaff** [comment]
In your [official guide](https://docs.aws.amazon.com/apprunner/latest/dg/apprunner-guide.pdf) it says this, which is completely unhelpful because **there are no logs**

![Screen Shot 2022-04-22 at 8 18 43 AM](https://user-images.githubusercontent.com/1238684/164772090-b2befdba-b5bc-4ff1-9a63-34433ab8780f.png)

**Comment by McLeroy** [comment]
I couldn't use App Runner because of this :(

**Comment by abba-logico-srl** [comment]
Same problem here :(

**Comment by michael-newsrx** [comment]
This morning I can't resume or create any services. :-(

Nothing logged in either case as to why.

**Comment by vladnicula** [comment]
same problem here...

**Comment by snnles** [comment]
Hi everyone,

We are working on providing more actionable logs for create service workflow. You will see incremental improvements in the 'Deployment Logs' and 'Event Logs' over the time.

Meanwhile, If you have any idea to share around how we can make logging and troubleshooting in App Runner better and what information you would like to see in 'Deployment Logs' and 'Event Logs', please share your ideas/feedback here.

Thanks!

**Comment by simllll** [comment]
Experience something simliar right now, 
"[AppRunner] Failed to deploy your application source code." but no reason why, is this also part of the improvements around logs?

**Comment by simllll** [comment]
@hariohmprasath thanks for your quick response,
e.g. arn:aws:apprunner:eu-west-1:770785616967:service/app-runner-git-deploy-service/aa6ff0e1cbc9414b84b9b104e0373d83
or arn:aws:apprunner:eu-west-1:770785616967:service/website-app-test/c017277194e04c08b05977db5dd7b11a

I guess both have the same issue though. Thanks a lot!

**Comment by hariohmprasath** [comment]
Hi @simllll,
I had a quick look at these services looks like both of them failed at build stage when running the build command. These build logs are available under `Deployment logs` section in the console. If you are fine you can share us a sample project without any proprietary code so we can try recreating this locally and help you fix this.

**Comment by simllll** [comment]
The deployment logs just show:
![image](https://user-images.githubusercontent.com/5757263/204602148-d062b8a5-f74b-412f-bda4-fc864371b1dc.png)
not really helpful.

Is there any possibility to get the build output?

**Comment by simllll** [comment]
Interesting, tried again just now, and now I got some output :-) I added to the npm install command the --ignore-scripts tag, not sure if this is related or some update just got rolled out. anyways, this makes things a lot easier (y)

**Comment by hariohmprasath** [comment]
> It really seems to be related to the "build" command, I changed it agian, and now I'm back with the "no output" issue.
> 
> build-command: `npm install --ignore-scripts && patch-package && npm run build`
> 
> I guess patch-package is failing, but somehow this leads to the behaviour that there is no output logged at all. I will try if it improves if I run a script instead.

Hi @simllll,
Yes based on our internal analysis its related to the build command, will you be able to share a sample project that you are using to run these tests? We would love to look at it and understand this better and improve the experience on our end. Thanks

**Comment by hariohmprasath** [comment]
> > It really seems to be related to the "build" command, I changed it agian, and now I'm back with the "no output" issue.
> > build-command: `npm install --ignore-scripts && patch-package && npm run build`
> > I guess patch-package is failing, but somehow this leads to the behaviour that there is no output logged at all. I will try if it improves if I run a script instead.
> 

Hi @simllll ,
Thanks for your patience. We are able to recreate this issue locally, looks like any errors related to build commands doesn't get transmitted to customer environment. We will work on a fix and update this thread in the upcoming weeks. Meanwhile you can workaround this problem by updating the build command to the pattern mentioned below and you will start seeing any command execution errors in the `Deployment logs`. 

```
(buildCommand) 2>&1
```

I was able to give it a try with an invalid command and was able to identify the error using the `Deployment logs` (refer below):
![Screen Shot 2022-11-29 at 3 06 36 PM](https://user-images.githubusercontent.com/2615325/204671447-0f56bd86-e3ab-42bf-b2cd-513800938589.png)

Hopefully this helps, let us know if you have any questions.

**Comment by simllll** [comment]
Thanks @hariohmprasath 
I succeeded with building now, at least according to the log (arn:aws:apprunner:eu-west-1:770785616967:service/website-app-deploy-test-envs/9b8c8ad1fa4a419f9960bac46f58b75a), last few log lines:


  | 2022-11-30T09:28:02.824+01:00 | [Build] [0mRemoving intermediate container 65911fcd9654
-- | -- | --
  | 2022-11-30T09:28:02.824+01:00 | [Build] ---> 46860de98696
  | 2022-11-30T09:28:02.824+01:00 | [Build] Step 5/5 : EXPOSE 18000
  | 2022-11-30T09:28:02.824+01:00 | [Build] ---> Running in a425b9b34785
  | 2022-11-30T09:28:02.824+01:00 | [Build] Removing intermediate container a425b9b34785
  | 2022-11-30T09:28:02.824+01:00 | [Build] ---> a7eb9cc657cf
  | 2022-11-30T09:28:02.824+01:00 | [Build] Successfully built a7eb9cc657cf
  | 2022-11-30T09:28:02.824+01:00 | [Build] Successfully tagged application-image:latest



looks actually good, but "Create service" is in state failed?

![image](https://user-images.githubusercontent.com/5757263/204747516-f350aa82-ef2a-4d0d-a85b-fa8613dc98e7.png)

Am I missing something?

**Comment by hariohmprasath** [comment]
> 

Hi @simllll,
I have moved this discussion to repost (https://repost.aws/questions/QUHRZ_kd9hR--kIUfAfrDdCg/app-runner-service-creation-failed) and also updated my observation to this service ARN

**Comment by asmfadholi** [comment]
Hi @hariohmprasath , I'm facing the same issue when deploying to app runner, and the error just said:
`[AppRunner] Failed to deploy your application source code.`

there is no detail provided. I just added `@sentry/nextjs` library in my nextjs project. Do you have insight about it?

Thank you.

**Comment by hariohmprasath** [comment]
Hi @asmfadholi,
I have moved this discussion to https://repost.aws/questions/QUYEjXJekeQ3O5Xz_YU5W1qw/app-runner-failed-to-deploy. Can you provide the serviceARN, part of your response? We can look into this and share some insights. Thanks

**Comment by asmfadholi** [comment]
@hariohmprasath I added arn to comment in aws discussion. Thank you.

**Comment by hariohmprasath** [comment]
Hi, @asmfadholi, 

Looks like you have run into a service bug because of a conflict in the NodeJS versions. We will prioritize this and work on a fix soon and keep you posted on the progress. As a temporary workaround you can install nodeJS 14 part of the `pre-build` phase and use that as a default runtime for building your code. This can be done by creating a configuration based service (using `apprunner.yaml`) instead of API based one.

Here is a sample `apprunner.yaml` file that you can use with the suggested fix:

```yaml
version: 1.0
runtime: nodejs14
build:
  commands:
    pre-build:
      - n 14.18.3
    build:      
      - <<build_command>>
run: 
  command: <<start_command>>
  network:
    port: <<port>>
    env: APP_PORT
```

Give it a try and let us know if you are still blocked on this. Thanks

**Comment by asmfadholi** [comment]
@hariohmprasath Thank you for the feedback. I was implementing your suggestion, but I still got an error. :(

**Comment by RichardGrant-93** [comment]
> ment logs. The last visible logs are

Well this didn't age well did it.

**Comment by RichardGrant-93** [comment]
> Hi everyone,
> 
> We are working on providing more actionable logs for create service workflow. You will see incremental improvements in the 'Deployment Logs' and 'Event Logs' over the time.
> 
> Meanwhile, If you have any idea to share around how we can make logging and troubleshooting in App Runner better and what information you would like to see in 'Deployment Logs' and 'Event Logs', please share your ideas/feedback here.
> 
> Thanks!

Well this didn't age well did it.

**Comment by januszm** [comment]
Not to mention that logs are flipped upside down, newest at the top, which completely breaks most of the multiline messages

**Comment by hariohmprasath** [comment]
> Not to mention that logs are flipped upside down, newest at the top, which completely breaks most of the multiline messages

Hi @januszm 
A follow up question: Did the service creation fail in your case? If so can you please create a post in repost.aws with the 'serviceARN', screenshot of your logs and share us the post URL? We will be happy to look into this and help you out. Thanks.

**Comment by januszm** [comment]
oh yeah it did fail. I'm actually still unable to launch a Rails application on AppRunner using "Source code repository".
What is repost.aws?

**Comment by hariohmprasath** [comment]
> oh yeah it did fail. I'm actually still unable to launch a Rails application on AppRunner using "Source code repository".
> 
> 

Can you share the *serviceARN* and if possible a sample ruby on rails app that you used in this test so we can look into this and get back to you. Thanks

**Comment by januszm** [comment]
I'll share it with our aws support representatives

**Comment by BahodiRajabov** [comment]
Me too
arn:aws:apprunner:eu-central-1:121334784369:service/tridmo_front/832792c9c49d4e10a9f27d368e498e09

**Comment by NachoMoneyNow** [comment]
Hey - I got this issue too.  everything was working - (deployments from a push worked, build scripts worked).
I also test the nextjs site before pushing for deployment using **npm run build**.

I made some updates to the code, did my **npm run build - it came out clean**, then I committed and pushed.  (**nothing else was changed - build scripts are still the same**).

the issue starts after _"[AppRunner] Starting source code build"_
then I get _"[AppRunner] Deployment failed. For details, see service logs."_

So, I have have no idea where the **"service logs"** are....

Here is our Service ARN:
arn:aws:apprunner:us-east-1:125560917055:service/vulcanPortalDev2/663fff4ff8184a12a6dd7f069a34b835

here is the date and time that it happened:
_05-04-2023 02:58:34 PM (GMT +8 hours)_ 

**Can a helpful AWS App Runner team member look up the Service log for this error and give me more error details?**  

Thanks

**Comment by mstoyanovv** [comment]
More than a year and still no decent way to check logs about why a deploy is failing...

**Comment by tnurmi82** [comment]
Yeah, me too. I'm new to App Runner. I have a .NET Core app which works fine locally, also works fine if ran with Docker locally, but I can't get it to work in App Runner.  "Create Failed".

```
06-02-2023 02:02:31 PM [AppRunner] Your application stopped or failed to start. See logs for more information.  Container exit code: 1
06-02-2023 02:00:25 PM [AppRunner] Performing health check on path '/' and port '443'.
06-02-2023 02:00:15 PM [AppRunner] Provisioning instances and deploying image for publicly accessible service.
06-02-2023 02:00:03 PM [AppRunner] Successfully built source code.
```

"See logs for more information" What logs? Where? I can't find any additional logs anywhere, and I'm stuck here in "Create Failed". Tried googling like crazy but cannot find any help.

**Comment by runlaiyang** [comment]
Hi @NachoMoneyNow, thanks for reaching out!  

I had a quick look at this service, while the latest deployments are successful. The failed ones were at build stage when running the build command. 

Build logs are available in CloudWatch
Log group > `/aws/apprunner/service-name/service-id/service`
Log steam > `deployment/operation-id`
Within `deployment/operation-id`, each log message starts with prefix `[Build]` is the output of your own build scripts.

Hope this would help on your analysis on build phases during deployment.

**Comment by vnedelchev** [comment]
The issue I encountered was related to building a Docker image from a Mac with an M1 chip. To overcome this problem, I used the following command: "docker build --platform linux/amd64 -t <image name> .". This command specifically targets the Linux/amd64 architecture, ensuring compatibility with AWS App Runner. The issue has been acknowledged by AWS, and you can find more information about it in this GitHub thread: https://github.com/aws/apprunner-roadmap/issues/1. I hope this helps.

**Comment by TimSon777** [comment]
Hi! No updates here? I am using .net 6, asp.net core. The application stopped deploying, the aws configuration was not changed, a similar aws configuration for a different environment (more strict) works. Old versions also do not work, everything works on a different environment and locally

Logs (nothing about health): 
```
Service status is set to RUNNING.
Service update failed. For details, see service logs.
Provisioning instances and deploying image.
Successfully copied the image from ECR.
Service status is set to OPERATION_IN_PROGRESS.
Service update started.
```

**Comment by Frans06** [comment]
Same here, deploys were working amazing until yesterday no changes on infra. this is my service. arn:aws:apprunner:us-east-2:986761589464:service/platform-staging-platform/13ee82df0d1948368d9f769b30f3d935 

same logs from here.

> Hi! No updates here? I am using .net 6, asp.net core. The application stopped deploying, the aws configuration was not changed, a similar aws configuration for a different environment (more strict) works. Old versions also do not work, everything works on a different environment and locally
> 
> Logs (nothing about health):
> 
> ```
> Service status is set to RUNNING.
> Service update failed. For details, see service logs.
> Provisioning instances and deploying image.
> Successfully copied the image from ECR.
> Service status is set to OPERATION_IN_PROGRESS.
> Service update started.
> ```

**Comment by Frans06** [comment]
For all this people including @TimSon777 , I solve my problem just pausing the app runner service, wait a time to let AWS kill the current Node and then run service again. After that your service will run fine. (I think It is what happens behind the wires).

**Comment by enriquinho-catalan** [comment]
I received this error through Pulumi due to an IAM problem. Try to look through Cloudtrail to see if anything is out-of-sorts. I got equally unuseful logs.

**Comment by TimSon777** [comment]
On the same day, the price of apprunner skyrocketed, almost 5 times. Why is that? @Frans06, do you also?

**Comment by TimSon777** [comment]
@enriquinho-catalan, I studied everything there, because I thought that someone had broken it. I didn't see anything out of the ordinary.

**Comment by timxyz** [comment]
In the past I've had my services silently fail because I had a `.python-version` file committed to the source repository, which AppRunner didn't like. Sharing that in case it helps someone in the same situation.

Unfortunately that is not the reason why I'm currently seeing the issue. My service runs fine when I manually specify the build configuration, but not when I use an equivalent `apprunner.yaml` file, and I have no idea why because there are no relevant logs.

**Comment by szandara** [comment]
I chose AWS App Runner because it was supposed to help me quickly deploy my webapp. I waste hours trying to understand what's going on without luck. Would have been easier to just use EC2.

This is all I have


```
2023-11-01T14:22:47.639+01:00 | [AppRunner] Starting to build your application source code.
```

**Comment by cpc-dev-contactpro** [comment]
@szandara I was stuck on that same error a few times, but eventually I got through it. Now I'm creating my App Runner services using CloudFormation to avoid any manual mistakes and it's working nicely. My 2 cents:

1. If that error appears, give it a minute, then refresh to see if any Application Logs have appeared. For some reason there is a significant delay sometimes, until the App Log appears. In my case I had defined HTTP as the method for App Runner to monitor if my app is alive (root path, port 80). But my app has auth, and root path was returning 401 Unauthorized (silly mistake). Now I ensure that HTTP GET / always returns 200 OK.
2. On another occasion, my apprunner.yaml was incorrect. This time I got no Application Log (because App Runner couldn't build the app), and the vague Deployment Log didn't indicate where the issue was. I looked at my past commits, and realized I had recently made an indentation error in my YAML file.
3. When I later switched from using CLI only to using CloudFormation, I once again got the same error. This time the issue was in my CloudFormation template itself. I was providing incorrect set of parameters for App Runner, and it was causing App Runner not to read my apprunner.yaml file. Again, same vague error on CloudFormation side. Eventually I figured out my mistake in the CF template.

But after all the hurdles, I'm currently very happy. I think that overall App Runner is saving me time and therefore money. I'm now able to sping up new App Runner services very quickly. Let's see if my opinion changes during following months, after I've used App Runner in production for a while.

But I still agree AWS could improve the docs, error messages and logging overall. I must have wasted more than 20 hours on these issues. Better logs would have saved a lot of time.

**Comment by tnurmi82** [comment]
@szandara  I was stuck on that same error a few times, but eventually I got through it. Now I'm creating my App Runner services using CloudFormation to avoid any manual mistakes and it's working nicely. My 2 cents:

If that error appears, give it a minute, then refresh to see if any Application Logs have appeared. For some reason there is a significant delay sometimes, until the App Log appears. In my case I had defined HTTP as the method for App Runner to monitor if my app is alive (root path, port 80). But my app has auth, and root path was returning 401 Unauthorized (silly mistake). Now I ensure that HTTP GET / always returns 200 OK.
On another occasion, my apprunner.yaml was incorrect. This time I got no Application Log (because App Runner couldn't build the app), and the vague Deployment Log didn't indicate where the issue was. I looked at my past commits, and realized I had recently made an indentation error in my YAML file.
When I later switched from using CLI only to using CloudFormation, I once again got the same error. This time the issue was in my CloudFormation template itself. I was providing incorrect set of parameters for App Runner, and it was causing App Runner not to read my apprunner.yaml file. Again, same vague error on CloudFormation side. Eventually I figured out my mistake in the CF template.
But after all the hurdles, I'm currently very happy. I think that overall App Runner is saving me time and therefore money. I'm now able to sping up new App Runner services very quickly. Let's see if my opinion changes during following months, after I've used App Runner in production for a while.

But I still agree AWS could improve the docs, error messages and logging overall. I must have wasted more than 20 hours on these issues. Better logs would have saved a lot of time.

**Comment by runlaiyang** [comment]
> I chose AWS App Runner because it was supposed to help me quickly deploy my webapp. I waste hours trying to understand what's going on without luck. Would have been easier to just use EC2.
> 
> This is all I have
> 
> ```
> 2023-11-01T14:22:47.639+01:00 | [AppRunner] Starting to build your application source code.
> ```

Hi @szandara
Thanks for reaching out, can you let us know the service ARN so we can provide you with some details around the service creation failure?

**Comment by szandara** [comment]
@runlaiyang I ended up using AWS Amplify, which seems to do very similar things. There I could find the log that I needed. In fact it was a simple missing dependency.

**Comment by neidiom** [comment]
I had similar problems and when I switched from HTTP to TCP health checks the deployment passed.

**Comment by hariohmprasath** [comment]
@neidiom, TCP health checks only validates whether the port is open before marking the service creation as success. But in case of HTTP we perform checks on the specified health check URL to get back a valid 200 response before updating the status. 

Here are few things you can try:
1. Can you add some loggers to application health check endpoints to verify whether these endpoints are actually getting invoked as part of App Runner health checks
2. How long does your application take to startup? Is it more than 1 minute?

**Comment by mitchboulay** [comment]
Adding a known working vpc connector to connect to RDS fails and the service rollsback. I used the public image to test the connection (public.ecr.aws/f0j5z9b5/netcat:latest ), locally it works and without the change, my container starts right up. The moment I add the VPC connector, it fails with the same message above, then rolls back the changes.

**Comment by LipatovAlexander** [comment]
I have another issue with updating app runner. It says "App Runner instance stopped due to an internal system error. Contact AWS support for assistance." whenever I'm trying to update image version.

![image](https://github.com/aws/apprunner-roadmap/assets/77479370/179e098b-8bba-45be-ad0c-adac3ebc65a5)

`arn:aws:apprunner:us-east-1:347648581624:service/tilly-production/55566502705444b4a46895be197c15a8`

**Comment by LipatovAlexander** [comment]
@runlaiyang Can you help me, please? ⬆️

**Comment by runlaiyang** [comment]
> @runlaiyang Can you help me, please? ⬆️

Hi @LipatovAlexander , thanks for reaching out. We are seeing instance terminated with Exit Code 139. This indicates a segmentation error – a memory violation, caused by a container trying to access a memory location to which it does not have access. Can you check application log see if any error from application code?

Please open a support ticket with the ARN of the service you were trying to update if you need more assistance.

**Comment by LipatovAlexander** [comment]
> > @runlaiyang Can you help me, please? ⬆️
> 
> Hi @LipatovAlexander , thanks for reaching out. We are seeing instance terminated with Exit Code 139. This indicates a segmentation error – a memory violation, caused by a container trying to access a memory location to which it does not have access. Can you check application log see if any error from application code?
> 
> Please open a support ticket with the ARN of the service you were trying to update if you need more assistance.

Yes, I found it. It was an error when starting my application. But it said about internal system error (on the AWS side, I assume) and it confused me.

Thank you!

**Comment by dbmikus** [comment]
I'm getting the same kind of problem. Service creation fails, and I get no logs or indication why. I only see a log like:

> Service creation failed. For details, see service logs.
> Service status is set to CREATE_FAILED.

The service ARN is arn:aws:apprunner:us-east-2:851725343267:service/GRPCServerStaging/99948daadf3f4257a7f3e33e0894cd1d

Relevant info:

- this is a gRPC server that exposes port 50051
- health check is a TCP check
- I am accessing Systems Manager Parameters, which I believe I have configured IAM for correctly

Can you help me debug why it won't start?

Is it a problem loading in the image? Or loading in the Systems Manager Parameters? Thanks.

**Comment by runlaiyang** [comment]
> I'm getting the same kind of problem. Service creation fails, and I get no logs or indication why. I only see a log like:
> 
> > Service creation failed. For details, see service logs.
> > Service status is set to CREATE_FAILED.
> 
> The service ARN is arn:aws:apprunner:us-east-2:851725343267:service/GRPCServerStaging/99948daadf3f4257a7f3e33e0894cd1d
> 
> Relevant info:
> 
>     * this is a gRPC server that exposes port 50051
> 
>     * health check is a TCP check
> 
>     * I am accessing Systems Manager Parameters, which I believe I have configured IAM for correctly
> 
> 
> Can you help me debug why it won't start?
> 
> Is it a problem loading in the image? Or loading in the Systems Manager Parameters? Thanks.

Hi @dbmikus , thanks for contacting us!
Look like the initial create service was failed on exit code 1, and multiple attempts to rebuild service failed with above log you saw. 
From our side, we see action `kms:DescribeKey` was access denied with the provided KMS key.  
Can you check if KMS policy has been changed?  Or you can try configure key policy with full access to the KMS key.

Please feel free to contact AWS support if you need more assistance. Thanks

**Comment by amine-ecom** [comment]
I faced the same issue with an error "Failed to run build command" but when I used this build command: "npm install --ignore-scripts" it worked fine for me

**Comment by BinyominCohen** [comment]
Unsubscribe

**Comment by crohr** [comment]
Using a docker image, I get a completely useless deployment error, with no application logs at all:

```
02-14-2024 03:49:35 PM [AppRunner] Deployment with ID : e4749593e76a4097a2e4da553d208c35 started. Triggering event : SERVICE_CREATE
02-14-2024 03:49:35 PM [AppRunner] Deployment Artifact: [Repo Type: ECR-Public], [Image URL: public.ecr.aws/c5h5o9k1/runs-on/runs-on], [Image Tag: v1.6.1-dev]
02-14-2024 03:49:55 PM [AppRunner] Pulling image public.ecr.aws/c5h5o9k1/runs-on/runs-on from ECR-Public repository.
02-14-2024 03:49:57 PM [AppRunner] Successfully pulled your application image from ECR.
02-14-2024 03:50:06 PM [AppRunner] Deployment with ID : e4749593e76a4097a2e4da553d208c35 failed.
```

The image is public, so you can try it if you want.

**Comment by hariohmprasath** [comment]
Thanks for connecting with us on this. Can you provide the port number used to setup the service? Did you had a chance to look at the container logs? You can view them using the application logs tab or directly inside cloudwatch

**Comment by crohr** [comment]
There were no application logs at all. It's port 80, which I believe is set by apprunner. There were no changes on that side compared to the previous version of the container.

**Comment by crohr** [comment]
@hariohmprasath if it helps, it looks like this is only when creating a new AppRunner service. If I use the same container on an existing service, it works fine.

**Comment by hariohmprasath** [comment]
> There were no application logs at all. It's port 80, which I believe is set by apprunner. There were no changes on that side compared to the previous version of the container.

Hi @crohr,
Looks like you are using the wrong port, the application is running on `3000` so when i used it as part of App Runner service configuration the service got created successfully with this URL - https://ieapemimb7.us-east-1.awsapprunner.com/

**Comment by 1oglop1** [comment]
<img width="1165" alt="image" src="https://github.com/aws/apprunner-roadmap/assets/7480694/bd6f7c14-8ac9-43ee-8f8c-22081862bb54">
but there are not logs ...

**Comment by rhythmshandilya** [comment]
Hi, this 2024 already. Anyone taking this up?

**Comment by don-attilio** [comment]
I am sharing how I solved the error I was hitting, even if what happened behind the curtains is beyond my understanding.
In my setup I had a public endpoint with egress traffic routed through a VPC. The source was an image set on a private ECR repository.
The `/application` log group was not created at all, while the service logs were not helpful:

```
2024-03-18T12:14:45.169 [AppRunner] Starting to deploy your application image.
2024-03-18T12:19:06.432 [AppRunner] Failed to deploy your application image.
```

By looking at other working services in my account, I realised that AppRunner couldn't pull the image from ECR.
The VPC connector had two private subnets, whose route table had an entry pointing to the NAT Gateway.
For reasons that are still obscure to me, removing the NAT Gateway from the route table was the only way to have the service up and running.
When AppRunner connects to ECR, pushes the logs to CloudWatch, or reads parameters from Parameter Store/Secrets Manager, the traffic is not routed through the VPC. Since my application required no connectivity except some services I had VPC endpoints for, removing the NAT gateway seemed like a reasonable move.
I struggle to find a valid explanation for the workaround above, but hopefully it will save some time to others.

**Comment by neidiom** [comment]
Any progress on this task?

**Comment by deshazer** [comment]
I was trying to deploy a Node.js app and ran into this same problem. App Runner would fail during my build command with no logs to tell me why. Hopefully this might help someone...

I have found that I never receive any logs for the "build" command. However, I do receive logs for the "start" command when configuring App Runner from the console. So, my debugging process was:

1. Move all build commands to the start command instead. Reminder: you can't use `&&` in a build command. Instead, define a new script in your package.json file like `"apprunner": "npm run install && npm run build && npm run start"`. Then, make your start command `npm run apprunner`. Hopefully that makes sense.
2. Look at the Application Logs on AWS to see everything that happened during the "start" process
3. Fix issues and then move everything that works successfully back to the "build" command

As a final heads up, if you depend on environment variables to run your build command, you need to find a different way to insert those during build. The environment variables you define in the App Runner console do not appear to be available during the "build" phase (only the "start" phase).

The big takeaway is: AWS SHOULD PUT APP RUNNER BUILD LOGS INTO THE APPLICATION LOGS!

**Comment by kleenkanteen** [comment]
I am deploying a simple nextjs app. I tried using app runner and it failed. Afetr that, I did what deshazer above suggested about creating a `apprunner` script and running that in the start command. It still failed.

Also, @deshazer you have a typo in your code, the `npm run install` should be `npm install`, as your code looks for a `install` script. So the the proper code is `"apprunner": "npm install && npm run build && npm run start"`. For the apprunner script you can also just write `"apprunner": "npm run build && npm run start". `, and in the start command, write `npm install`.

After trying what deshazer wrote and it still failing, the logs are still useless. I looked at deployment logs and application logs. Deployment logs end with this:

```
08-14-2024 07:50:34 PM [PostBuild]  ---> 83c7607fbe4b
08-14-2024 07:50:34 PM [PostBuild] Successfully built 83c7607fbe4b
08-14-2024 07:52:23 PM [AppRunner] Successfully built your application source code.
08-14-2024 07:59:17 PM [AppRunner] Failed to deploy your application source code.
```

why did it fail? no helpful info.

applicaiton logs end with this:
```
08-14-2024 06:54:08 PM info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
08-14-2024 07:53:34 PM > keepsake@0.1.0 apprunner
08-14-2024 07:53:34 PM > npm run build && npm run start
08-14-2024 07:53:35 PM > keepsake@0.1.0 build
08-14-2024 07:53:35 PM > next build
08-14-2024 07:53:36 PM Attention: Next.js now collects completely anonymous telemetry regarding usage.
08-14-2024 07:53:36 PM This information is used to shape Next.js' roadmap and prioritize features.
08-14-2024 07:53:36 PM You can learn more, including how to opt-out if you'd not like to participate in this anonymous program, by visiting the following URL:
08-14-2024 07:53:36 PM https://nextjs.org/telemetry
08-14-2024 07:53:36 PM    ▲ Next.js 14.1.0
08-14-2024 07:53:36 PM    Linting and checking validity of types ...
08-14-2024 07:54:28 PM ./src/components/application/shared/files/FileTypeIcons.tsx
08-14-2024 07:54:28 PM 150:20  Warning: Using `<img>` could result in slower LCP and higher bandwidth. Consider using `<Image />` from `next/image` to automatically optimize images. This may incur additional usage or cost from your provider. See: https://nextjs.org/docs/messages/no-img-element  @next/next/no-img-element
08-14-2024 07:54:28 PM info  - Need to disable some ESLint rules? Learn more here: https://nextjs.org/docs/basic-features/eslint#disabling-rules
```

I don't see any errors, why did it stop here?

I tried deploying my app on render.com instead and I got i working with 1 minute of config time. compare this to 3 hours of bashing my head with app runner and it still is failing.

It's a crime how bad the DX is with app runner. Similar exp with ECS. I am avoiding aws like the plague from now on.

edit: The issue was that I wasn't passing in environment variables needed during build, casuing it to fail when building. I moved to GCP cloud run and worked great.

**Comment by BreezyPod** [comment]
Not an App Runner expert, but I wonder if this is a byproduct of a pull-based architecture where the agent has a bunch of logic and needs to connect to the control plane to communicate some failure state. If that connection can't be established then there is missing info. If that is the case I am not sure how App Runner is in a position to improve this experience, would be interested in learning more!

**Comment by davidgilbertson** [comment]
I came here looking for a solution to "Failed to deploy your application source code." but instead I've become convinced that App Runner is not really safe to use (as in, I wouldn't use it for clients, only to have to tell them in a year that I can't deploy that urgent update and can't explain why).

As it stands, I can't even get the "Getting started" tutorial deployed with the sample AWS code and have lost a day trying due to slow deploys and lack of log-based feedback.

**Comment by kleenkanteen** [comment]
@davidgilbertson Use GCP Cloud Run, I switched to that and it's so much better. Logs actually get printed. render.com has even better dx, but a tad bit more expensive.

**Comment by LewisW** [comment]
GCP cloud run is considerably better, it's just a shame you need a global load balancer ($18/month) to assign a custom domain. I think you're right though, AppRunner isn't fit for service and I'll stop wasting my time trying to get it to work, back to GCP I go.

**Comment by vinduajamesh** [comment]
Wow, it's been 2 years and the logging is still vague. Come on, people are moving away from AWS to something like GCP Cloud  Run and you guys aren't doing anything about it.

I'm not gonna lie, AWS App Runner is great and easy to set up, but the ridiculous and vague logging when something goes wrong is preventing people from using it.

**Comment by jonjack** [comment]
It is unbelievable how many rough edges AWS has for a leading cloud platform. I wasted 5 hours on App Runner because of a silly mistake that pretty much every other runtime environment I have ever used would have logged. This ticket has sat here for over 3 years now with seemingly no progress and its not exactly a complex fix. Pretty shocking - no wonder the market share of Azure and GCP are growing.

**Comment by obryen** [comment]
Exactly 3 years later and I am here to report the same issue, I am attempting to deploy an image from ECR of a nestjs app to app runner but it fails after this message ` [AppRunner] Provisioning instances and deploying image for publicly accessible service.` . The issue for me is very much the same as what others on here have reported, something is failing at the application layer or even the control plane layer but all I get to see is what is highlighted in the screenshot, This is incredibly annoying , especially after 3 whole years, I have also checked the `application logs` section and it is very much empty. I will add that I have attempted to run the image from ECR locally and found and fixed some underlying DB access issues which I initially thought was what was preventing the app from running on `App Runner` as well but now that all that is ironed outm  my local tests are working fine but app runner is still showing this message. 

<img width="1397" alt="Image" src="https://github.com/user-attachments/assets/5154e6ea-bd41-4edc-b1f3-49a2acaa6d76" />

**Comment by januszm** [comment]
@obryen  try Elastic Beanstalk with Docker, it's a much more mature, stable, complete platform for web applications, that is also easy to configure and requires almost zero maintenance. It's actively maintained and regularly receives updates.

**Comment by neidiom** [comment]
@obryen ECS on EC2 is also great. I switched from AppRunner to ECS.

**Comment by obryen** [comment]
<img width="824" alt="Image" src="https://github.com/user-attachments/assets/9822811b-13aa-4c1e-9b49-7ce60d8b93e0" />

> Exactly 3 years later and I am here to report the same issue, I am attempting to deploy an image from ECR of a nestjs app to app runner but it fails after this message ` [AppRunner] Provisioning instances and deploying image for publicly accessible service.` . The issue for me is very much the same as what others on here have reported, something is failing at the application layer or even the control plane layer but all I get to see is what is highlighted in the screenshot, This is incredibly annoying , especially after 3 whole years, I have also checked the `application logs` section and it is very much empty. I will add that I have attempted to run the image from ECR locally and found and fixed some underlying DB access issues which I initially thought was what was preventing the app from running on `App Runner` as well but now that all that is ironed outm my local tests are working fine but app runner is still showing this message.
> 
> 

[Update] I think aws fixed this issue with no logs on the `App runner` level, the latest `create-service` command just displayed actual logs! Amazing how easy things became from here. Thank you for the update!

**Comment by OmarKhater-sws** [comment]
I still have the same issue after 3 years from this issue!

**Comment by j-q-in-berlin** [comment]
Any news here? This is really frustrating when everything is running smoothly locally, but failing without any logs on app runner.

**Comment by Deepak-at-Bridgly** [comment]
This is so frustrating! 3 years and still no logs in deployment logs : (

**Comment by Ben-CA** [comment]
Still no update on this?

**Comment by j-q-in-berlin** [comment]
> [@obryen](https://github.com/obryen) ECS on EC2 is also great. I switched from AppRunner to ECS.

I wonder if AWS is just making App Runner shitty on purpose to force people to switch to a more expensive product. This is so frustrating.

**Comment by tnurmi82** [comment]
I don't work for AWS, but I would be interested to read in more detail about the issues people are still facing.

I've been a very happy App Runner user now for over 2 years, using the managed .NET 6 runtime:
1. Ensure release build always works locally without problems.
2. If there is a deployment problem, first check Deployment Log (duh).
3. If Deployment log is vague and doesn't tell me what went wrong, there is likely an issue happening in my code. I should wait 1-2min and then check Application logs.

Initially I was too fast and didn't realize the "Application logs" can have a small delay until more lines are shown.
Other than that, no problemos with App Runner, at least with the managed .NET 6.

However, now that the managed .NET 6 runtime is being deprecated, I'll need to switch from managed runtime to ECR-based solution. Let's see how that goes, and if the logs somehow suck more in that scenario.

**Comment by j-q-in-berlin** [comment]
The thing is that sometimes the deployment fails and you only have deployment logs and NO application logs whatsover. So I had the app running fine locally but failing on App Runner without any meaningful logs to look at.
(But I am not working with .NET)

**Comment by neidiom** [comment]
In my experience there are always logs in CloudWatch, just the order is not what you expect. Try reading last 2 log streams.

**Comment by j-q-in-berlin** [comment]
Believe me. Not always, and if you look at other comments here and on reddit for example you will see that there are many people with the same problem.

**Comment by neidiom** [comment]
I switched to ECS. Now I can SSH into the EC2 instance and use fast rolling deployments instead of blue-green, which are damn slow.

**Comment by davidhughhenrymack** [comment]
Please please give us helpful logs. this is unusable.

```
11-18-2025 02:35:23 PM [AppRunner] Creating pipeline for automatic deployments.
11-18-2025 02:35:23 PM [AppRunner] Successfully created pipeline for automatic deployments.
11-18-2025 02:35:24 PM [AppRunner] Pulling source code from GITHUB Repository ( https://github.com/... ).
11-18-2025 02:35:47 PM [AppRunner] Successfully pulled your application source code.
11-18-2025 02:36:03 PM [AppRunner] Reading apprunner.yaml config file.
11-18-2025 02:36:11 PM [AppRunner] Deployment with ID : 1d2c7588a67742f08b606c526639e953 failed.
```

Nothing more anywhere to help. None of the echo stmts in my pre-run show up.

**Comment by tnurmi82** [comment]
@davidhughhenrymack would you like to share your apprunner.yml? Maybe we could spot something there.

But it does indeed seem that AWS wants us to move away from using the managed runtimes & "pull source from GitHub"-mode, most obviously with .NET. Too bad, because I really liked the feature.

Having said that - I have now switched to the ECR option. After a few hours of getting it to work, I do see the benefits. First of all, it is now much easier to install any prerequisites I need (in my case, some dependencies for Puppeteer & Chrome). I can just do apt-get in Dockerfile (before it was 'yam install' in my apprunner.yml, and it took me hours to get working). Deployment is overall faster now. Compared to before, I have of course full control of the pipeline now. I use GitHub actions and at least so far I'm easily managing with the free tier. At the end of my pipeline my new container version is pushed to ECR, and then App Runner just deploys it. And even though I had no real problems with App Runner's logs before, it is way easier to follow pipeline logs in GitHub. And now the only deployment problems on App Runner side are likely about IAM Roles and log pretty clearly states what is wrong.

So.... I guess I do see why AWS is making this move. They would have to invest quite a lot to improve App Runner on this section, and partially compete with GitLab CI and GitHub Actions. Would they get any profit out of it - Likely their math says "not enough".

**Comment by Ben-CA** [comment]
If AWS teams are strictly thinking that customers will only use another AWS service, then they might not see the need to invest in AppRunner. But they should consider that developers will start to favor Google Cloud Run, Azure App Service, Vercel, etc. if they feel that AppRunner is a dead end at AWS.

Most developers don't want to fight with DevOps and maintaining servers - they want something that "just works".

**The lack of attention goes deeper than poor build logs:**
Why is AppRunner only offered in some regions?
Why does AppRunner not offer integrated email notifications on build (like Elastic Beanstalk and Amplify, etc. do).
Why does AppRunner's "built-in" options only include PHP 8.1 - reaching end of support shortly - and not offer PHP 8.4? (or even 8.2 or 8.3)
